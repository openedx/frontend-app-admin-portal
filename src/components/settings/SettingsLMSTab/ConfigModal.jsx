import React from 'react';
import PropTypes from 'prop-types';
import { ModalDialog, ActionRow, Button } from '@edx/paragon';

const MODAL_TITLE = 'Do you want to save your work?';
const MODAL_TEXT = 'Your changes will be lost without saving.';

// will have to pass in individual saveDraft method and config when
// drafting is allowed
const ConfigModal = ({ isOpen, close, onClick }) => (
  <ModalDialog
    title="Cancel Modal"
    isOpen={isOpen}
    onClose={close}
    variant="default"
  >
    <ModalDialog.Header>
      <ModalDialog.Title>
        {MODAL_TITLE}
      </ModalDialog.Title>
    </ModalDialog.Header>
    <ModalDialog.Body>
      {MODAL_TEXT}
    </ModalDialog.Body>
    <ModalDialog.Footer>
      <ActionRow>
        <Button onClick={() => onClick('')} variant="outline-primary">
          Exit without saving
        </Button>
        <Button variant="primary">
          Save
        </Button>
      </ActionRow>
    </ModalDialog.Footer>
  </ModalDialog>
);

ConfigModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
};
export default ConfigModal;
