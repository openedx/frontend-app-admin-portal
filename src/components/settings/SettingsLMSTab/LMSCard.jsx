import React from 'react';
import PropTypes from 'prop-types';
import { Card, Image } from '@edx/paragon';
import { getLMSIcon } from '../../../utils';

const LMSCard = ({ LMSType, onClick }) => (
  <Card isClickable tabIndex="0" className="p-2.5 lms-card" onClick={() => onClick(LMSType)}>
    <Card.Body>
      <h3 className="text-center">
        <Image className="lms-icon" src={getLMSIcon(LMSType)} />
        <span className="ml-2">{LMSType}</span>
      </h3>
    </Card.Body>
  </Card>
);

LMSCard.propTypes = {
  LMSType: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};
export default LMSCard;
