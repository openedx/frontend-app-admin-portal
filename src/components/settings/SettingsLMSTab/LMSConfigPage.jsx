import React from 'react';
import PropTypes from 'prop-types';
import { Image } from '@edx/paragon';
import { connect } from 'react-redux';
import { logError } from '@edx/frontend-platform/logging';
import { getLMSIcon } from '../../../utils';

import {
  BLACKBOARD_TYPE,
  CANVAS_TYPE,
  CORNERSTONE_TYPE,
  DEGREED_TYPE,
  MOODLE_TYPE,
  SAP_TYPE,
} from '../data/constants';
import BlackboardConfig from './LMSConfigs/BlackboardConfig';
import CanvasConfig from './LMSConfigs/CanvasConfig';
import CornerstoneConfig from './LMSConfigs/CornerstoneConfig';
import DegreedConfig from './LMSConfigs/DegreedConfig';
import MoodleConfig from './LMSConfigs/MoodleConfig';
import SAPConfig from './LMSConfigs/SAPConfig';

export function buttonBool(config) {
  let returnVal = true;
  Object.values(config).forEach(value => {
    if (!value) {
      returnVal = false;
    }
  });
  return returnVal;
}

export function handleErrors(error) {
  const errorMsg = error.message || error.response?.status <= 300
    ? error.message
    : JSON.stringify(error.response.data);
  logError(errorMsg);
  return errorMsg;
}

const LMSConfigPage = ({ LMSType, onClick, enterpriseId }) => (
  <span>
    <h3 className="mt-4.5 mb-3.5">
      <Image className="lms-icon" src={getLMSIcon(LMSType)} />
      <span className="ml-2">Connect {LMSType}</span>
    </h3>
    {LMSType === BLACKBOARD_TYPE && (
    <BlackboardConfig id={enterpriseId} onClick={onClick} />
    )}
    {LMSType === CANVAS_TYPE && (
    <CanvasConfig id={enterpriseId} onClick={onClick} />
    )}
    {LMSType === CORNERSTONE_TYPE && (
    <CornerstoneConfig id={enterpriseId} onClick={onClick} />
    )}
    {LMSType === DEGREED_TYPE && (
    <DegreedConfig id={enterpriseId} onClick={onClick} />
    )}
    {LMSType === MOODLE_TYPE && (
    <MoodleConfig id={enterpriseId} onClick={onClick} />
    )}
    {LMSType === SAP_TYPE && (
    <SAPConfig id={enterpriseId} onClick={onClick} />
    )}
  </span>
);

const mapStateToProps = (state) => ({
  enterpriseId: state.portalConfiguration.enterpriseId,
});

LMSConfigPage.propTypes = {
  enterpriseId: PropTypes.string.isRequired,
  LMSType: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default connect(mapStateToProps)(LMSConfigPage);
