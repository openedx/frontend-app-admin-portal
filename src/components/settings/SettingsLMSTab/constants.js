export const BLACKBOARD_ICON = 'https://uthsc.edu/tlc/images/bblogoc.png';
export const CANVAS_ICON = 'https://w7.pngwing.com/pngs/883/509/png-transparent-iphone-learning-management-system-canvas-lms-canvas-electronics-canvas-mobile-phones-thumbnail.png';
export const CORNERSTONE_ICON = 'https://pbs.twimg.com/profile_images/1455186964785758212/iA1nUdl2_400x400.jpg';
export const DEGREED_ICON = 'https://workforceedtech.org/wp-content/uploads/2019/03/degreed_logo_resized.png';
export const MOODLE_ICON = 'https://pittcc.edu/wp-content/uploads/2018/04/moodle-logo.png';
export const SAP_ICON = 'https://upload.wikimedia.org/wikipedia/commons/thumb/5/59/SAP_2011_logo.svg/1600px-SAP_2011_logo.svg.png';
