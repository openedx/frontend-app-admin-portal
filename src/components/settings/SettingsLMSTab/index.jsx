import React, { useEffect, useState } from 'react';
import {
  Alert, Button, Hyperlink, CardGrid, Toast,
} from '@edx/paragon';
import { Link } from 'react-router-dom';
import { Add, Info } from '@edx/paragon/icons';
import { logError } from '@edx/frontend-platform/logging';
import PropTypes from 'prop-types';

import { camelCaseDictArray } from '../../../utils';
import LMSCard from './LMSCard';
import LMSConfigPage from './LMSConfigPage';
import ExistingLMSCardDeck from './ExistingLMSCardDeck';
import {
  BLACKBOARD_TYPE,
  CANVAS_TYPE,
  CORNERSTONE_TYPE,
  DEGREED2_TYPE,
  HELP_CENTER_LINK,
  MOODLE_TYPE,
  SAP_TYPE,
  SUCCESS_LABEL,
  DELETE_SUCCESS_LABEL,
  TOGGLE_SUCCESS_LABEL,
} from '../data/constants';
import LmsApiService from '../../../data/services/LmsApiService';

const SUBMIT_TOAST_MESSAGE = 'Configuration was submitted successfully.';
const TOGGLE_TOAST_MESSAGE = 'Configuration was toggled successfully.';
const DELETE_TOAST_MESSAGE = 'Configuration was successfully removed.';

export default function SettingsLMSTab({
  enterpriseId,
  enterpriseSlug,
  enableSamlConfigurationScreen,
  identityProvider,
}) {
  const [config, setConfig] = useState();
  const [showToast, setShowToast] = useState(false);

  const [existingConfigsData, setExistingConfigsData] = useState({});
  const [configsExist, setConfigsExist] = useState(false);
  const [showNewConfigButtons, setShowNewConfigButtons] = useState(true);

  const [existingConfigFormData, setExistingConfigFormData] = useState({});
  const [toastMessage, setToastMessage] = useState();
  const [displayNeedsSSOAlert, setDisplayNeedsSSOAlert] = useState(false);

  // onClick function for existing config cards' edit action
  const editExistingConfig = (configData, configType) => {
    // Set the form data to the card's
    setExistingConfigFormData(configData);
    // Set the config type to the card's type
    setConfig(configType);
    // Hide the create new configs button
    setShowNewConfigButtons(true);
    // Since the user is editing, hide the existing config cards
    setConfigsExist(false);
  };

  const fetchExistingConfigs = () => {
    const options = { enterprise_customer: enterpriseId };
    LmsApiService.fetchEnterpriseCustomerIntegrationConfigs(options)
      .then((response) => {
        // If the enterprise has existing configs
        if (response.data.length !== 0) {
          // Save all existing configs
          setExistingConfigsData(camelCaseDictArray(response.data));
          // toggle the existing configs bool
          setConfigsExist(true);
          // Hide the create cards and show the create button
          setShowNewConfigButtons(false);
        }
      })
      .catch((error) => {
        // We can ignore errors here as user will se the banner in the next page refresh.
        logError(error);
      });
  };

  const onClick = (input) => {
    // Either we're creating a new config (a create config card was clicked), or we're navigating
    // back to the landing state from a form (submit or cancel was hit on the forms). In both cases,
    // we want to clear existing config form data.
    setExistingConfigFormData({});

    // If either the user has submit or canceled
    if (input === '' || [SUCCESS_LABEL, DELETE_SUCCESS_LABEL, TOGGLE_SUCCESS_LABEL].includes(input)) {
      // Re-fetch existing configs to get newly created ones
      fetchExistingConfigs();
    }
    // If the user submitted
    if (input === SUCCESS_LABEL) {
      // show the toast and reset the config state
      setShowToast(true);
      setConfig('');
      setToastMessage(SUBMIT_TOAST_MESSAGE);
    } else if (input === TOGGLE_SUCCESS_LABEL) {
      setShowToast(true);
      setConfig('');
      setToastMessage(TOGGLE_TOAST_MESSAGE);
    } else if (input === DELETE_SUCCESS_LABEL) {
      setShowToast(true);
      setConfig('');
      setToastMessage(DELETE_TOAST_MESSAGE);
    } else {
      // Otherwise the user has clicked a create card and we need to set existing config bool to
      // false and set the config type to the card that was clicked type
      setConfigsExist(false);
      setConfig(input);
    }
  };

  // onClick function for the show create cards button
  const showCreateConfigCards = () => {
    setShowNewConfigButtons(true);
  };

  useEffect(() => {
    // On load fetch potential existing configs
    fetchExistingConfigs();
  }, []);

  useEffect(() => {
    // Check if the customer needs an idp config
    setDisplayNeedsSSOAlert(enableSamlConfigurationScreen && !identityProvider);
  }, [enableSamlConfigurationScreen, identityProvider]);

  return (
    <div>
      <div className="d-flex">
        <h2 className="py-2">Learning Management System Configuration</h2>
        <Hyperlink
          destination={HELP_CENTER_LINK}
          className="btn btn-outline-primary ml-auto my-2"
          target="_blank"
        >
          Help Center
        </Hyperlink>
      </div>
      <p>
        Enabling a learning management system for your edX account allows quick
        access to the catalog
      </p>
      {displayNeedsSSOAlert && (
        <Alert
          className="mr-6 sso-alert-modal-margin"
          variant="danger"
          icon={Info}
          actions={[
            <Link to={`/${enterpriseSlug}/admin/settings/sso`}>
              <Button>Configure SSO</Button>
            </Link>,
          ]}
        >
          <Alert.Heading>No SSO configured</Alert.Heading>
          <p>
            At least one Single Sign On configuration is needed to create an LMS configuration
          </p>
        </Alert>
      )}
      {configsExist && (
        <span>
          <h4 className="mt-5 mb-4">Existing configurations</h4>
          <ExistingLMSCardDeck
            configData={existingConfigsData}
            editExistingConfig={editExistingConfig}
            enterpriseCustomerUuid={enterpriseId}
            onClick={onClick}
          />
        </span>
      )}
      {!showNewConfigButtons && (
        <Button
          varient="primary"
          className="mr-6"
          iconBefore={Add}
          size="lg"
          block
          disabled={displayNeedsSSOAlert}
          onClick={showCreateConfigCards}
        >
          New configuration
        </Button>
      )}
      {showNewConfigButtons && !config && (
        <span>
          <h4 className="mt-5">New configurations</h4>
          <p className="mb-4">Click on a card to start a new configuration</p>

          <CardGrid
            columnSizes={{
              xs: 6,
              s: 5,
              m: 4,
              l: 4,
              xl: 3,
            }}
          >
            <LMSCard LMSType={SAP_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
            <LMSCard LMSType={MOODLE_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
            <LMSCard LMSType={CORNERSTONE_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
            <LMSCard LMSType={CANVAS_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
            <LMSCard LMSType={DEGREED2_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
            <LMSCard LMSType={BLACKBOARD_TYPE} disabled={displayNeedsSSOAlert} onClick={onClick} />
          </CardGrid>
        </span>
      )}
      {config && (
        <span>
          <LMSConfigPage LMSType={config} onClick={onClick} existingConfigFormData={existingConfigFormData} />
        </span>
      )}
      {toastMessage && (
        <Toast
          onClose={() => setShowToast(false)}
          show={showToast}
        >
          {toastMessage}
        </Toast>
      )}
    </div>
  );
}

SettingsLMSTab.propTypes = {
  enterpriseId: PropTypes.string.isRequired,
  enterpriseSlug: PropTypes.string.isRequired,
  enableSamlConfigurationScreen: PropTypes.bool.isRequired,
  identityProvider: PropTypes.string.isRequired,
};
