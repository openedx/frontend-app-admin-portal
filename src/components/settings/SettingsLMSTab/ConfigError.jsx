import React from 'react';
import PropTypes from 'prop-types';
import { AlertModal, ActionRow, Button } from '@edx/paragon';
import { HELP_CENTER_LINK } from '../data/constants';

const cardText = 'We were unable to process your request to submit a new LMS configuration. Please try submitting again or contact support for help.';

const ConfigError = ({ isOpen, close, submit }) => (
  <AlertModal
    title="Something went wrong"
    isOpen={isOpen}
    onClose={close}
    footerNode={(
      <ActionRow>
        <Button
          href={HELP_CENTER_LINK}
          className="ml-auto my-2"
          rel="noopener noreferrer"
          target="_blank"
          variant="tertiary"
        >
          Help Center
        </Button>
        <Button variant="primary" onClick={submit}>Submit</Button>
      </ActionRow>
    )}
  >
    <p>
      {cardText}
    </p>
  </AlertModal>
);

ConfigError.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
};
export default ConfigError;
