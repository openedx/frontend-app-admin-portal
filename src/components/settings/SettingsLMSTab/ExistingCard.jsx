import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  ActionRow, AlertModal, Badge, Button, Card, Dropdown, Icon, IconButton, Image, OverlayTrigger, Popover,
} from '@edx/paragon';
import { MoreVert } from '@edx/paragon/icons';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { features } from '../../../config';
import { channelMapping } from '../../../utils';
import handleErrors from '../utils';
import { TOGGLE_SUCCESS_LABEL, DELETE_SUCCESS_LABEL } from '../data/constants';

const errorToggleModalText = 'We were unable to toggle your configuration. Please try submitting again or contact support for help.';
const errorDeleteModalText = 'We were unable to delete your configuration. Please try removing again or contact support for help.';
const INCOMPLETE = 'incomplete';
const ACTIVE = 'active';
const INACTIVE = 'inactive';

const ExistingCard = ({
  config,
  editExistingConfig,
  enterpriseCustomerUuid,
  onClick,
  openError,
  openReport,
  setReportConfig,
  setErrorModalText,
  getStatus,
}) => {
  const [showDeleteModal, setShowDeleteModal] = useState(false);
  const isEdxStaff = getAuthenticatedUser().administrator;

  const openModalButton = () => {
    setReportConfig(config);
    openReport();
  };

  const toggleConfig = async (id, channelType, toggle) => {
    const configOptions = {
      active: toggle,
      enterprise_customer: enterpriseCustomerUuid,
    };
    let err;
    try {
      await channelMapping[channelType].update(configOptions, id);
    } catch (error) {
      err = handleErrors(error);
    }
    if (err) {
      setErrorModalText(errorToggleModalText);
      openError();
    } else {
      onClick(TOGGLE_SUCCESS_LABEL);
    }
  };

  const deleteConfig = async (id, channelType) => {
    let err;
    try {
      await channelMapping[channelType].delete(id);
    } catch (error) {
      err = handleErrors(error);
    }
    if (err) {
      setErrorModalText(errorDeleteModalText);
      openError();
    } else {
      onClick(DELETE_SUCCESS_LABEL);
      setShowDeleteModal(false);
    }
  };

  const handleClickDelete = (isInactive) => {
    if (isInactive) {
      setShowDeleteModal(true);
    } else {
      deleteConfig(config.id, config.channelCode);
    }
  };

  const numIncorrectFields = (fields) => {
    const { missing } = fields[0];
    const { incorrect } = fields[1];
    let totalNum = 0;
    if (missing.includes('refresh_token')) {
      totalNum = -1;
    }
    totalNum = totalNum + missing.length + incorrect.length;
    let strongText;
    if (totalNum === 0) {
      return <span>Please <strong>authorize your LMS</strong> to submit this form.</span>;
    }
    if (totalNum === 1) {
      strongText = `${totalNum} field`;
    } else { strongText = `${totalNum} fields`; }
    return <span>Amend <strong>{strongText}</strong> to submit this form.</span>;
  };

  const getCardButton = () => {
    switch (getStatus(config)) {
      case ACTIVE:
        if (isEdxStaff && features.FEATURE_INTEGRATION_REPORTING) {
          return <Button variant="outline-primary" onClick={() => openModalButton(config)}>View sync history</Button>;
        }
        return null;
      case INCOMPLETE:
        return <Button variant="outline-primary" onClick={() => editExistingConfig(config, config.channelCode)}>Configure</Button>;
      case INACTIVE:
        return <Button variant="outline-primary" onClick={() => toggleConfig(config.id, config.channelCode, true)}>Enable</Button>;
      default:
        return null;
    }
  };

  const isActive = getStatus(config) === ACTIVE;
  const isInactive = getStatus(config) === INACTIVE;
  const isIncomplete = getStatus(config) === INCOMPLETE;

  return (
    <>
      {/* TODO: Figure out how to get rid of scroll bar */}
      <AlertModal
        title="Delete integration?"
        isOpen={showDeleteModal}
        onClose={() => setShowDeleteModal(false)}
        hasCloseButton
        footerNode={(
          <ActionRow>
            <Button
              variant="tertiary"
              data-testid="cancel-delete-config"
              onClick={() => setShowDeleteModal(false)}
            >
              Cancel
            </Button>
            <Button
              variant="danger"
              data-testid="confirm-delete-config"
              onClick={() => deleteConfig(config.id, config.channelCode)}
            >
              Delete
            </Button>
          </ActionRow>
        )}
      >
        <p>
          Are you sure you want to delete this learning platform integration?
          Once deleted, any saved integration data will be lost.
        </p>
      </AlertModal>
      <Card
        tabIndex="0"
        className="p-3 existing-lms-card"
        key={config.channelCode + config.id}
      >
        <Card.Header
          className="lms-card-content"
          actions={(
            <Dropdown>
              <Dropdown.Toggle
                id="dropdown-toggle-with-iconbutton"
                data-testid={`existing-lms-config-card-dropdown-${config.id}`}
                as={IconButton}
                src={MoreVert}
                iconAs={Icon}
                variant="primary"
                alt="Actions dropdown"
              />
              <Dropdown.Menu>
                {(isInactive && isEdxStaff && features.FEATURE_INTEGRATION_REPORTING) && (
                  <div className="d-flex">
                    <Dropdown.Item
                      onClick={() => openModalButton(config)}
                      data-testid="dropdown-sync-history-item"
                    >
                      View sync history
                    </Dropdown.Item>
                  </div>
                )}
                {isActive && (
                  <div className="d-flex">
                    <Dropdown.Item
                      onClick={() => toggleConfig(config.id, config.channelCode, false)}
                      data-testid="dropdown-disable-item"
                    >
                      Disable
                    </Dropdown.Item>
                  </div>
                )}
                {(isInactive || isIncomplete) && (
                  <div className="d-flex">
                    <Dropdown.Item
                      // Ask before deleting an inactive project
                      onClick={() => handleClickDelete(isInactive)}
                      data-testid="dropdown-delete-item"
                    >
                      Delete
                    </Dropdown.Item>
                  </div>
                )}
                {!isIncomplete && (
                  <div className="d-flex">
                    <Dropdown.Item
                      onClick={() => editExistingConfig(config, config.channelCode)}
                    >
                      Configure
                    </Dropdown.Item>
                  </div>
                )}
              </Dropdown.Menu>
            </Dropdown>
        )}
          title={(
            <div className="ml-1 d-flex">
              <Image
                className="lms-icon mr-2"
                src={channelMapping[config.channelCode].icon}
              />
              <div className="lms-card-title-overflow">
                <span>{config.displayName}</span>
              </div>
              <h3 className="ml-2">
                {/* Only incomplete badges will show hover */}
                <OverlayTrigger
                  trigger={['hover', 'focus']}
                  key={`${config.channelCode + config.id}hover`}
                  placement="top"
                  overlay={(
                    <Popover className="popover-positioned-top" id="inc-popover">
                      <Popover.Title as="h5">Next Steps</Popover.Title>
                      <Popover.Content>
                        {numIncorrectFields(config.isValid)}
                      </Popover.Content>
                    </Popover>
            )}
                >
                  <span>
                    {isIncomplete && (
                    <Badge variant="light">Incomplete</Badge>
                    )}
                  </span>
                </OverlayTrigger>
                {isInactive && (
                <Badge variant="light">Disabled</Badge>
                )}
              </h3>
            </div>
        )}
        />
        <Card.Footer className="pt-2 pb-2 justify-content-end">
          {getCardButton()}
        </Card.Footer>
      </Card>
    </>
  );
};

ExistingCard.propTypes = {
  config: PropTypes.shape({
    active: PropTypes.bool,
    isValid: PropTypes.arrayOf(
      PropTypes.shape({
        missing: PropTypes.arrayOf(PropTypes.string),
        incorrect: PropTypes.arrayOf(PropTypes.string),
      }),
    ),
    channelCode: PropTypes.string,
    id: PropTypes.number,
    displayName: PropTypes.string,
  }).isRequired,
  editExistingConfig: PropTypes.func.isRequired,
  enterpriseCustomerUuid: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  openError: PropTypes.func.isRequired,
  openReport: PropTypes.func.isRequired,
  setReportConfig: PropTypes.func.isRequired,
  setErrorModalText: PropTypes.func.isRequired,
  getStatus: PropTypes.func.isRequired,
};

export default ExistingCard;
