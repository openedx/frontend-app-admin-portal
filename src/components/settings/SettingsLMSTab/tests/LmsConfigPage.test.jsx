import React from 'react';
import {
  fireEvent, screen,
} from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

import { renderWithRouter } from '../../../test/testUtils';
import {
  BLACKBOARD_TYPE,
  CANVAS_TYPE,
  CORNERSTONE_TYPE,
  DEGREED_TYPE,
  MOODLE_TYPE,
  SAP_TYPE,
} from '../../data/constants';

import SettingsLMSTab from '../index';
import { buttonBool } from '../LMSConfigPage';

const initialState = {
  portalConfiguration: {
    enterpriseId: 'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee',
  },
};

const draftConfig = {
  field1: '',
  field2: 'test',
  field3: 'test',
};

const completeConfig = {
  field1: 'test',
  field2: 'test',
  field3: 'test',
};

const mockStore = configureMockStore([thunk]);

const SettingsLMSWrapper = () => (
  <Provider store={mockStore({ ...initialState })}>
    <SettingsLMSTab />
  </Provider>
);

describe('<SettingsLMSTab />', () => {
  it('Renders with all LMS cards present', () => {
    renderWithRouter(<SettingsLMSWrapper />);
    expect(screen.queryByText(BLACKBOARD_TYPE)).toBeTruthy();
    expect(screen.queryByText(CANVAS_TYPE)).toBeTruthy();
    expect(screen.queryByText(CORNERSTONE_TYPE)).toBeTruthy();
    expect(screen.queryByText(DEGREED_TYPE)).toBeTruthy();
    expect(screen.queryByText(MOODLE_TYPE)).toBeTruthy();
    expect(screen.queryByText(SAP_TYPE)).toBeTruthy();
  });
  test('Blackboard card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const blackboardCard = screen.getByText(BLACKBOARD_TYPE);
    fireEvent.click(blackboardCard);
    expect(screen.queryByText('Connect Blackboard')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect Blackboard')).toBeFalsy();
  });
  test('Canvas card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const canvasCard = screen.getByText(CANVAS_TYPE);
    fireEvent.click(canvasCard);
    expect(screen.queryByText('Connect Canvas')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect Canvas')).toBeFalsy();
  });
  test('Cornerstone card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const cornerstoneCard = screen.getByText(CORNERSTONE_TYPE);
    fireEvent.click(cornerstoneCard);
    expect(screen.queryByText('Connect Cornerstone')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect Cornerstone')).toBeFalsy();
  });
  test('Degreed card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const degreedCard = screen.getByText(DEGREED_TYPE);
    fireEvent.click(degreedCard);
    expect(screen.queryByText('Connect Degreed')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect Degreed')).toBeFalsy();
  });
  test('Moodle card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const moodleCard = screen.getByText(MOODLE_TYPE);
    fireEvent.click(moodleCard);
    expect(screen.queryByText('Connect Moodle')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect Moodle')).toBeFalsy();
  });
  test('SAP card cancel flow', async () => {
    renderWithRouter(<SettingsLMSWrapper />);
    const sapCard = screen.getByText(SAP_TYPE);
    fireEvent.click(sapCard);
    expect(screen.queryByText('Connect SAP')).toBeTruthy();
    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Do you want to save your work?')).toBeTruthy();
    const exitButton = screen.getByText('Exit without saving');
    fireEvent.click(exitButton);
    expect(screen.queryByText('Connect SAP')).toBeFalsy();
  });
  test('test button text', () => {
    expect(!buttonBool(draftConfig));
    expect(buttonBool(completeConfig));
  });
});
