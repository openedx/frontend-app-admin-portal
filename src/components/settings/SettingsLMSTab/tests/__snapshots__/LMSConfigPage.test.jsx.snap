// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsLMSTab /> renders correctly 1`] = `
<div>
  <div
    className="d-flex"
  >
    <h2
      className="py-2"
    >
      Learning Management System Configuration
    </h2>
    <a
      className="ml-auto my-2 btn btn-outline-primary"
      href="https://business-support.edx.org/hc/en-us/categories/360000368453-Integrations"
      onClick={[Function]}
      onKeyDown={[Function]}
      rel="noopener noreferrer"
      target="_blank"
    >
      Help Center
    </a>
  </div>
  <p>
    Enabling a learning management system for your edX account allows quick access to the catalog
  </p>
  <span>
    <h4
      className="mt-5"
    >
      New configurations
    </h4>
    <p
      className="mb-4"
    >
      Click on a card to start a new configuration
    </p>
    <div
      className="pgn__card-grid"
    >
      <div
        className="row"
      >
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="icon/mock/path"
                />
                <span
                  className="ml-2"
                >
                  SAP
                </span>
              </h3>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="test-file-stub"
                />
                <span
                  className="ml-2"
                >
                  Moodle
                </span>
              </h3>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="test-file-stub"
                />
                <span
                  className="ml-2"
                >
                  Cornerstone
                </span>
              </h3>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="icon/mock/path"
                />
                <span
                  className="ml-2"
                >
                  Canvas
                </span>
              </h3>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="test-file-stub"
                />
                <span
                  className="ml-2"
                >
                  Degreed
                </span>
              </h3>
            </div>
          </div>
        </div>
        <div
          className="col-xl-3 col-6"
          l={4}
          m={4}
          s={5}
        >
          <div
            className="lms-card card"
            isClickable={true}
            onClick={[Function]}
            tabIndex="0"
          >
            <div
              className="card-body"
            >
              <h3
                className="text-center"
              >
                <img
                  className="lms-icon"
                  src="icon/mock/path"
                />
                <span
                  className="ml-2"
                >
                  Blackboard
                </span>
              </h3>
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;
